src/binaryutils.cpp

```diff
@@ -0,0 +1,284 @@
+    std::vector<T> nchwVector(rows * cols * channels);
+    for (int y = 0; y < rows; ++y) {
+        for (int x = 0; x < cols; ++x) {
+            for (int c = 0; c < channels; ++c) {
+                nchwVector[c * (rows * cols) + y * cols + x] = reinterpret_cast<const T*>(nhwcVector)[y * channels * cols + x * channels + c];
+            }
+        }
+    }
+    return std::move(nchwVector);
+}
+
+cv::Mat convertStringValToMat(const std::string& stringVal) {
+    std::vector<unsigned char> data(stringVal.begin(), stringVal.end());
+    cv::Mat dataMat(data, true);
+
+    return cv::imdecode(dataMat, cv::IMREAD_UNCHANGED);
+}
+
+Status convertPrecision(const cv::Mat& src, cv::Mat& dst, const InferenceEngine::Precision requestedPrecision) {
+    int type = getMatTypeFromTensorPrecision(requestedPrecision);
+    if (type == -1) {
+        return StatusCode::INVALID_PRECISION;
+    }
+
+    src.convertTo(dst, type);
+    return StatusCode::OK;
+}
+
+bool resizeNeeded(const cv::Mat& image, const std::shared_ptr<TensorInfo>& tensorInfo) {
+    if (tensorInfo->getLayout() == InferenceEngine::Layout::NCHW) {
+        int cols = tensorInfo->getShape()[3];
+        int rows = tensorInfo->getShape()[2];
+        if (cols != image.cols || rows != image.rows) {
+            return true;
+        }
+    } else if (tensorInfo->getLayout() == InferenceEngine::Layout::NHWC) {
+        int cols = tensorInfo->getShape()[2];
+        int rows = tensorInfo->getShape()[1];
+        if (cols != image.cols || rows != image.rows) {
+            return true;
+        }
+    }
+    return false;
+}
+
+Status resizeMat(const cv::Mat& src, cv::Mat& dst, const std::shared_ptr<TensorInfo>& tensorInfo) {
+    if (tensorInfo->getLayout() == InferenceEngine::Layout::NCHW) {
+        int cols = tensorInfo->getShape()[3];
+        int rows = tensorInfo->getShape()[2];
+        cv::resize(src, dst, cv::Size(cols, rows));
+
+        return StatusCode::OK;
+    } else if (tensorInfo->getLayout() == InferenceEngine::Layout::NHWC) {
+        int cols = tensorInfo->getShape()[2];
+        int rows = tensorInfo->getShape()[1];
+        cv::resize(src, dst, cv::Size(cols, rows));
+
+        return StatusCode::OK;
+    }
+    return StatusCode::UNSUPPORTED_LAYOUT;
+}
+
+Status validateNumberOfChannels(const std::shared_ptr<TensorInfo>& tensorInfo,
+    const cv::Mat input) {
+    // Network and input must have the same number of shape dimensions.
+    if (tensorInfo->getLayout() == InferenceEngine::Layout::NCHW) {
+        if ((unsigned int)(input.channels()) != tensorInfo->getShape()[1]) {
+            SPDLOG_DEBUG("Binary sent to input: {} has invalid number of channels. Expected: {} Actual: {}", tensorInfo->getMappedName(), tensorInfo->getShape()[1], input.channels());
+            return StatusCode::INVALID_NO_OF_CHANNELS;
+        }
+    } else if (tensorInfo->getLayout() == InferenceEngine::Layout::NHWC) {
```

Comment:
My latest findings are that when model has NHWC layout it still reports old shape so it is at position 1.